File: back_velora/src/main/java/com/velora/catalog/CatalogController.java
package com.velora.catalog;

import java.net.URI;
import java.util.List;
import java.util.UUID;

import org.springframework.core.io.Resource;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import com.velora.catalog.asset.CatalogAssetService;
import com.velora.catalog.dto.CategoryRequest;
import com.velora.catalog.dto.CategoryResponse;
import com.velora.catalog.dto.ImageRequest;
import com.velora.catalog.dto.ImageResponse;
import com.velora.catalog.dto.ProductRequest;
import com.velora.catalog.dto.ProductResponse;
import com.velora.catalog.dto.VariantRequest;
import com.velora.catalog.dto.VariantResponse;
import com.velora.catalog.image.ProductImagePurpose;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/catalog")
public class CatalogController {

    private final CatalogService catalogService;
    private final CatalogAssetService catalogAssetService;

    public CatalogController(
            CatalogService catalogService,
            CatalogAssetService catalogAssetService
    ) {
        this.catalogService = catalogService;
        this.catalogAssetService = catalogAssetService;
    }

    @GetMapping("/categories")
    public List<CategoryResponse> categories() {
        return catalogService.listPublicCategories();
    }

    @GetMapping("/products")
    public List<ProductResponse> products() {
        return catalogService.listPublicProducts();
    }

    @GetMapping("/products/{id}")
    public ProductResponse product(@PathVariable UUID id) {
        return catalogService.getPublicProduct(id);
    }

    @GetMapping("/products/{id}/variants")
    public List<VariantResponse> variants(@PathVariable UUID id) {
        return catalogService.listPublicVariants(id);
    }

    @GetMapping("/manage/categories")
    public List<CategoryResponse> managedCategories() {
        return catalogService.listManagedCategories();
    }

    @PostMapping("/manage/categories")
    public ResponseEntity<CategoryResponse> createCategory(
            @Valid @RequestBody CategoryRequest request
    ) {
        CategoryResponse created = catalogService.createCategory(request);

        return ResponseEntity
                .created(URI.create(
                        "/api/catalog/manage/categories/" + created.id()
                ))
                .body(created);
    }

    @PutMapping("/manage/categories/{id}")
    public CategoryResponse updateCategory(
            @PathVariable UUID id,
            @Valid @RequestBody CategoryRequest request
    ) {
        return catalogService.updateCategory(id, request);
    }

    @GetMapping("/manage/products")
    public List<ProductResponse> managedProducts() {
        return catalogService.listManagedProducts();
    }

    @GetMapping("/manage/products/{id}")
    public ProductResponse managedProduct(@PathVariable UUID id) {
        return catalogService.getManagedProduct(id);
    }

    @PostMapping("/manage/products")
    public ResponseEntity<ProductResponse> createProduct(
            @Valid @RequestBody ProductRequest request,
            @AuthenticationPrincipal Jwt jwt
    ) {
        ProductResponse created = catalogService.createProduct(
                request,
                actorId(jwt)
        );

        return ResponseEntity
                .created(URI.create(
                        "/api/catalog/manage/products/" + created.id()
                ))
                .body(created);
    }

    @PutMapping("/manage/products/{id}")
    public ProductResponse updateProduct(
            @PathVariable UUID id,
            @Valid @RequestBody ProductRequest request,
            @AuthenticationPrincipal Jwt jwt
    ) {
        return catalogService.updateProduct(
                id,
                request,
                actorId(jwt)
        );
    }

    @PostMapping("/manage/products/{productId}/variants")
    public ResponseEntity<VariantResponse> createVariant(
            @PathVariable UUID productId,
            @Valid @RequestBody VariantRequest request
    ) {
        VariantResponse created = catalogService.createVariant(
                productId,
                request
        );

        return ResponseEntity.ok(created);
    }

    @PutMapping("/manage/variants/{id}")
    public VariantResponse updateVariant(
            @PathVariable UUID id,
            @Valid @RequestBody VariantRequest request
    ) {
        return catalogService.updateVariant(id, request);
    }

    @PostMapping("/manage/products/{productId}/images")
    public ResponseEntity<ImageResponse> createImage(
            @PathVariable UUID productId,
            @Valid @RequestBody ImageRequest request
    ) {
        return ResponseEntity.ok(
                catalogService.createImage(productId, request)
        );
    }

    @PostMapping(
            value = "/manage/products/{productId}/images/upload",
            consumes = MediaType.MULTIPART_FORM_DATA_VALUE
    )
    public ResponseEntity<ImageResponse> uploadImage(
            @PathVariable UUID productId,
            @RequestParam(required = false) UUID variantId,
            @RequestParam(required = false) String altText,
            @RequestParam(defaultValue = "GALLERY")
            ProductImagePurpose purpose,
            @RequestParam(defaultValue = "0") Integer sortOrder,
            @RequestParam(defaultValue = "false") Boolean primary,
            @RequestParam("file") MultipartFile file
    ) {
        return ResponseEntity.ok(
                catalogAssetService.uploadProductImage(
                        productId,
                        variantId,
                        altText,
                        purpose,
                        sortOrder,
                        primary,
                        file
                )
        );
    }

    @GetMapping("/assets/{storageKey:.+}")
    public ResponseEntity<Resource> catalogAsset(
            @PathVariable String storageKey
    ) {
        var loaded = catalogAssetService.load(storageKey);

        return ResponseEntity.ok()
                .contentType(
                        MediaType.parseMediaType(
                                loaded.contentType()
                        )
                )
                .body(loaded.resource());
    }

    private UUID actorId(Jwt jwt) {
        return UUID.fromString(jwt.getSubject());
    }
}

File: apps/back/src/catalog/catalog.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
  Query,
  Req,
  UseGuards,
} from "@nestjs/common";

import {
  catalogCompanyQuerySchema,
  categoryRequestSchema,
  imageRequestSchema,
  productRequestSchema,
  variantRequestSchema,
  type CatalogCompanyQuery,
  type CategoryRequest,
  type CategoryResponse,
  type ImageRequest,
  type ImageResponse,
  type ProductRequest,
  type ProductResponse,
  type VariantRequest,
  type VariantResponse,
} from "@velora/contracts";

import {
  BearerAuthGuard,
} from "../auth/bearer-auth.guard.js";

import type {
  AuthPrincipal,
} from "../auth/security.js";

import {
  RequireRoles,
} from "../common/authz/roles.decorator.js";

import {
  RolesGuard,
} from "../common/authz/roles.guard.js";

import type {
  AuthenticatedRequest,
} from "../common/http/authenticated-request.js";

import {
  ApiHttpError,
} from "../common/http/api-http.error.js";

import {
  ZodValidationPipe,
} from "../common/http/zod-validation.pipe.js";

import {
  CatalogService,
} from "./catalog.service.js";

@Controller("api/catalog")
export class CatalogPublicController {
  constructor(
    private readonly catalog:
      CatalogService,
  ) {}

  @Get("categories")
  categories(
    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<CategoryResponse[]> {
    return this.catalog
      .listPublicCategories(
        query.companyId,
      );
  }

  @Get("products")
  products(
    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<ProductResponse[]> {
    return this.catalog
      .listPublicProducts(
        query.companyId,
      );
  }

  @Get("products/:id")
  product(
    @Param(
      "id",
      new ParseUUIDPipe(),
    )
    id: string,

    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<ProductResponse> {
    return this.catalog
      .getPublicProduct(
        query.companyId,
        id,
      );
  }

  @Get("products/:id/variants")
  variants(
    @Param(
      "id",
      new ParseUUIDPipe(),
    )
    id: string,

    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<VariantResponse[]> {
    return this.catalog
      .listPublicVariants(
        query.companyId,
        id,
      );
  }
}

@Controller("api/catalog/manage")
@UseGuards(
  BearerAuthGuard,
  RolesGuard,
)
@RequireRoles(
  "ADMIN",
  "STORE_MANAGER",
)
export class CatalogManageController {
  constructor(
    private readonly catalog:
      CatalogService,
  ) {}

  @Get("categories")
  categories(
    @Req()
    request:
      AuthenticatedRequest,

    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<CategoryResponse[]> {
    return this.catalog
      .listManagedCategories(
        this.principal(
          request,
        ),
        query.companyId,
      );
  }

  @Post("categories")
  createCategory(
    @Req()
    request:
      AuthenticatedRequest,

    @Body(
      new ZodValidationPipe(
        categoryRequestSchema,
      ),
    )
    body:
      CategoryRequest,
  ): Promise<CategoryResponse> {
    return this.catalog
      .createCategory(
        this.principal(
          request,
        ),
        body,
      );
  }

  @Put("categories/:id")
  updateCategory(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "id",
      new ParseUUIDPipe(),
    )
    id: string,

    @Body(
      new ZodValidationPipe(
        categoryRequestSchema,
      ),
    )
    body:
      CategoryRequest,
  ): Promise<CategoryResponse> {
    return this.catalog
      .updateCategory(
        this.principal(
          request,
        ),
        id,
        body,
      );
  }

  @Get("products")
  products(
    @Req()
    request:
      AuthenticatedRequest,

    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<ProductResponse[]> {
    return this.catalog
      .listManagedProducts(
        this.principal(
          request,
        ),
        query.companyId,
      );
  }

  @Get("products/:id")
  product(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "id",
      new ParseUUIDPipe(),
    )
    id: string,

    @Query(
      new ZodValidationPipe(
        catalogCompanyQuerySchema,
      ),
    )
    query:
      CatalogCompanyQuery,
  ): Promise<ProductResponse> {
    return this.catalog
      .getManagedProduct(
        this.principal(
          request,
        ),
        query.companyId,
        id,
      );
  }

  @Post("products")
  createProduct(
    @Req()
    request:
      AuthenticatedRequest,

    @Body(
      new ZodValidationPipe(
        productRequestSchema,
      ),
    )
    body:
      ProductRequest,
  ): Promise<ProductResponse> {
    return this.catalog
      .createProduct(
        this.principal(
          request,
        ),
        body,
      );
  }

  @Put("products/:id")
  updateProduct(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "id",
      new ParseUUIDPipe(),
    )
    id: string,

    @Body(
      new ZodValidationPipe(
        productRequestSchema,
      ),
    )
    body:
      ProductRequest,
  ): Promise<ProductResponse> {
    return this.catalog
      .updateProduct(
        this.principal(
          request,
        ),
        id,
        body,
      );
  }

  @Post("products/:productId/variants")
  @HttpCode(
    HttpStatus.OK,
  )
  createVariant(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "productId",
      new ParseUUIDPipe(),
    )
    productId: string,

    @Body(
      new ZodValidationPipe(
        variantRequestSchema,
      ),
    )
    body:
      VariantRequest,
  ): Promise<VariantResponse> {
    return this.catalog
      .createVariant(
        this.principal(
          request,
        ),
        productId,
        body,
      );
  }

  @Put("variants/:id")
  updateVariant(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "id",
      new ParseUUIDPipe(),
    )
    id: string,

    @Body(
      new ZodValidationPipe(
        variantRequestSchema,
      ),
    )
    body:
      VariantRequest,
  ): Promise<VariantResponse> {
    return this.catalog
      .updateVariant(
        this.principal(
          request,
        ),
        id,
        body,
      );
  }

  @Post("products/:productId/images")
  @HttpCode(
    HttpStatus.OK,
  )
  createImage(
    @Req()
    request:
      AuthenticatedRequest,

    @Param(
      "productId",
      new ParseUUIDPipe(),
    )
    productId: string,

    @Body(
      new ZodValidationPipe(
        imageRequestSchema,
      ),
    )
    body:
      ImageRequest,
  ): Promise<ImageResponse> {
    return this.catalog
      .createImage(
        this.principal(
          request,
        ),
        productId,
        body,
      );
  }

  private principal(
    request:
      AuthenticatedRequest,
  ): AuthPrincipal {
    const principal =
      request.authPrincipal;

    if (!principal) {
      throw new ApiHttpError(
        401,
        "No autenticado.",
      );
    }

    return principal;
  }
}
